cli/tools/vendor/mod.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use std::path::Path;
+use std::path::PathBuf;
+
+use deno_core::anyhow::bail;
+use deno_core::error::AnyError;
+use deno_core::resolve_url_or_path;
+use deno_runtime::permissions::Permissions;
+
+use crate::flags::VendorFlags;
+use crate::fs_util;
+use crate::lockfile;
+use crate::proc_state::ProcState;
+use crate::resolver::ImportMapResolver;
+use crate::resolver::JsxResolver;
+
+mod analyze;
+mod build;
+mod import_map;
+mod mappings;
+mod specifiers;
+#[cfg(test)]
+mod test;
+
+pub async fn vendor(ps: ProcState, flags: VendorFlags) -> Result<(), AnyError> {
+  let raw_output_dir = match &flags.output_path {
+    Some(output_path) => output_path.to_owned(),
+    None => PathBuf::from("vendor/"),
+  };
+  let output_dir = fs_util::resolve_from_cwd(&raw_output_dir)?;
+  validate_output_dir(&output_dir, &flags, &ps)?;
+  let graph = create_graph(&ps, &flags).await?;
+  let vendored_count =
+    build::build(&graph, &output_dir, &build::RealVendorEnvironment)?;
+
+  eprintln!(
+    r#"Vendored {} {} into {} directory.
+
+To use vendored modules, specify the `--import-map` flag when invoking deno subcommands:
+  deno run -A --import-map {} main.ts"#,
+    vendored_count,
+    if vendored_count == 1 {
+      "module"
+    } else {
+      "modules"
+    },
+    raw_output_dir.display(),
+    raw_output_dir.join("import_map.json").display(),
+  );
+
+  Ok(())
+}
+
+fn validate_output_dir(
+  output_dir: &Path,
+  flags: &VendorFlags,
+  ps: &ProcState,
+) -> Result<(), AnyError> {
+  if !flags.force && !is_dir_empty(output_dir)? {
+    bail!(concat!(
+      "Output directory was not empty. Please specify an empty directory or use ",
+      "--force to ignore this error and potentially overwrite its contents.",
+    ));
+  }
+
+  // check the import map
+  if let Some(import_map_path) = ps
+    .maybe_import_map
+    .as_ref()
+    .and_then(|m| m.base_url().to_file_path().ok())
+    .and_then(|p| fs_util::canonicalize_path(&p).ok())
+  {
+    // make the output directory in order to canonicalize it for the check below
+    std::fs::create_dir_all(&output_dir)?;
+    let output_dir = fs_util::canonicalize_path(output_dir)?;
```

Comment:
Would be great to add `.context()` here to tell users at what stage something went wrong, otherwise the errors will be like `PermissionDenied (error 13)` or something like this. Very hard to debug